cloud/src/recycler/recycler_service.cpp

```diff
@@ -52,6 +58,215 @@ RecyclerServiceImpl::RecyclerServiceImpl(std::shared_ptr<TxnKv> txn_kv, Recycler
+             }},
+            {"rowset",
+             [](InstanceRecycler& instance_recycler) {
+                 instance_recycler.scan_and_statistics_rowsets();
+             }},
+            {"abort_timeout_txn",
+             [](InstanceRecycler& instance_recycler) {
+                 instance_recycler.scan_and_statistics_abort_timeout_txn();
+             }},
+            {"expired_txn_label",
+             [](InstanceRecycler& instance_recycler) {
+                 instance_recycler.scan_and_statistics_expired_txn_label();
+             }},
+            {"version",
+             [](InstanceRecycler& instance_recycler) {
+                 instance_recycler.scan_and_statistics_versions();
+             }},
+            {"copy_job",
+             [](InstanceRecycler& instance_recycler) {
+                 instance_recycler.scan_and_statistics_copy_jobs();
+             }},
+            {"stage",
+             [](InstanceRecycler& instance_recycler) {
+                 instance_recycler.scan_and_statistics_stage();
+             }},
+            {"expired_stage_object", [](InstanceRecycler& instance_recycler) {
+                 instance_recycler.scan_and_statistics_expired_stage_objects();
+             }}};
+
+    std::vector<std::string> invalid_resource_types;
+    // check invalid resource type
+    std::for_each(request->resource_type().begin(), request->resource_type().end(),
+                  [&](const std::string& resource_type) {
+                      if (!resource_handlers.contains(resource_type)) {
+                          invalid_resource_types.emplace_back(resource_type);
+                      }
+                  });
+    if (!invalid_resource_types.empty()) {
+        code = MetaServiceCode::INVALID_ARGUMENT;
+        msg = fmt::format(
+                "invalid resource type(s): {}, valid resource_type have [{}]",
+                fmt::join(invalid_resource_types, ", "),
+                std::accumulate(resource_handlers.begin(), resource_handlers.end(), std::string(),
+                                [&](const std::string& acc, const auto& pair) {
+                                    return acc.empty() ? pair.first : acc + ", " + pair.first;
+                                }));
+        LOG_WARNING(msg);
+        return;
+    }
+
+    std::set<std::string> instance_ids;
+    std::vector<InstanceInfoPB> instances;
+    get_all_instances(txn_kv_.get(), instances);
+
+    // check if a string matches a pattern with '*' and '?' wildcards.
+    auto isMatch = [&](const std::string& targetString, const std::string& patternString) {
+        const char* target = targetString.c_str();
+        const char* pattern = patternString.c_str();
+        const char* lastStar = nullptr;    // Pointer to the last '*' in the pattern
+        const char* backtrackPos = target; // Position to backtrack to if we hit a mismatch
+
+        while (*target) {
+            if (*pattern == '?' || *pattern == *target) {
```

Comment:
we dont need `?`
this function is too complex.
 `resource_type == "*"` or instance_id == "*" 
is enough